radixdlt/src/integration/java/com/radixdlt/consensus/simulation/checks/SafetyCheck.java

```diff
@@ -17,34 +17,105 @@
 
 package com.radixdlt.consensus.simulation.checks;
 
+import com.radixdlt.consensus.Vertex;
+import com.radixdlt.consensus.View;
 import com.radixdlt.consensus.simulation.BFTCheck;
 import com.radixdlt.consensus.simulation.SimulatedBFTNetwork;
-import com.radixdlt.consensus.VertexStore;
+import com.radixdlt.crypto.ECKeyPair;
+import com.radixdlt.utils.Pair;
 import io.reactivex.rxjava3.core.Observable;
-import java.util.Arrays;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicReference;
 import java.util.stream.Collectors;
 
 /**
- * Checks that nodes are committed on the same vertex.
+ * Checks that nodes do not commit on conflicting vertices
  */
 public class SafetyCheck implements BFTCheck {
 
+	private static Observable<BFTCheckError> differentVerticesError(Vertex vertex, Vertex currentVertex) {
+		return Observable.just(
+			new BFTCheckError(
+				String.format("Different vertices [%s, %s] committed at same view: %s",
+					vertex,
+					currentVertex,
+					vertex.getView()
+				)
+			)
+		);
+	}
+
+	private static Observable<BFTCheckError> noParentError(Vertex vertex) {
+		return Observable.just(
+			new BFTCheckError(
+				String.format("Committed vertex %s has no parent", vertex)
+			)
+		);
+	}
+
+	private static Observable<BFTCheckError> badParentError(Vertex vertex, Vertex lastCommitted) {
+		return Observable.just(
+			new BFTCheckError(
+				String.format("Parent of vertex %s doesn't match last committed %s",
+					vertex,
+					lastCommitted
+				)
+			)
+		);
+	}
+
 	@Override
 	public Observable<BFTCheckError> check(SimulatedBFTNetwork network) {
-		return Observable.zip(
-			network.getNodes().stream()
-				.map(network::getVertexStore)
-				.map(VertexStore::lastCommittedVertex)
-				.collect(Collectors.toList()),
-				Arrays::stream
+		final Map<View, Vertex> committedVertices = new ConcurrentHashMap<>();
+		committedVertices.put(View.genesis(), network.getGenesisVertex());
+		final AtomicReference<View> highest = new AtomicReference<>(View.genesis());
+		final Map<ECKeyPair, Vertex> lastCommittedByNode = new ConcurrentHashMap<>();
+
+		return Observable.merge(
+			network.getNodes().stream().map(
+				node -> network.getVertexStore(node)
+					.lastCommittedVertex().map(v -> Pair.of(node, v)))
+				.collect(Collectors.toList())
 		)
-			.map(committedVertices -> committedVertices.distinct().collect(Collectors.toList()))
-			.concatMap(committedVertices -> {
-				if (committedVertices.size() != 1) {
-					return Observable.just(new BFTCheckError("Committed vertices don't match: " + committedVertices));
+			.flatMap(nodeAndVertex -> {
+				final ECKeyPair node = nodeAndVertex.getFirst();
```

Comment:
Pretty long anonymous function inside this `flatMap`, making it hard to follow the reactive code, maybe break out to helper method? :)

Reply:
Breaking it out I think will make it harder to follow since this check requires a bit of state (e.g. committedVertices map). Could break it out if there was a bigger refactor to enable moving some state to the class level but that would need to be in the future.